Add extended open multilingual wordnet reader

ExplorerFreda opened the pull request:
Related issue: https://github.com/nltk/nltk/issues/1611 
Related NLTK data issue: https://github.com/nltk/nltk_data/issues/179
Extended open multilingual wordnet: http://compling.hss.ntu.edu.sg/omw/summx.html

ExplorerFreda committed:
Add extended open multilingual wordnet reader

ExplorerFreda committed:
Merge branch 'nltk:develop' into develop

tomaarsen commented:
Hey! We use [`pre-commit`](https://pre-commit.com/) hooks to ensure code quality through our [`pre-commit config`](https://github.com/nltk/nltk/blob/develop/.pre-commit-config.yaml), which runs `pyupgrade`, `black`, `isort` and some other hooks. Using it is as simple as `pip install pre-commit` and then running `pre-commit install`. From then onwards, any time you commit to `https://github.com/ExplorerFreda/nltk`, it will run these hooks before committing. 
The first time you might not have any changes to turn into a commit, so you can do `pre-commit run --all` or try to create an empty commit, and it will format your changes. You can then push this to make the CI stop complaining.

tomaarsen commented on nltk/corpus/reader/extended_omw.py:
```diff
@@ -0,0 +1,141 @@
+
+import os
+from collections import defaultdict
+from IPython.core.pylabtools import configure_inline_support
```
This seems like an unused import. Should this be there?

tomaarsen commented on nltk/corpus/reader/extended_omw.py:
```diff
@@ -0,0 +1,141 @@
+from collections import defaultdict
+from IPython.core.pylabtools import configure_inline_support
+
+from traitlets.traitlets import default
```
This seems like an unused import. Should this be there?

tomaarsen commented on nltk/corpus/reader/extended_omw.py:
```diff
@@ -0,0 +1,141 @@
+# Natural Language Toolkit: Extened Open Multilingual WordNet Reader
```
```suggestion
# Natural Language Toolkit: Extended Open Multilingual WordNet Reader
```

tomaarsen commented on nltk/corpus/reader/extended_omw.py:
```diff
@@ -0,0 +1,141 @@
+        for prov in self.provenances[lang]:
+            with self._omw_reader.open(f"{prov}/wn-data-{lang}.tab") as fp:
+                self.custom_lemmas(fp, lang)
+                fp.close()
```
The `with` keyword already automatically closes any files whenever the `with` scope is exited, so this line is unnecessary.

tomaarsen commented on nltk/corpus/reader/extended_omw.py:
```diff
@@ -0,0 +1,141 @@
+        for prov in self.exomw_provenances[lang]:
+            with self._exomw_reader.open(f"{prov}/wn-{prov}-{lang}.tab") as fp:
+                self.custom_lemmas(fp, lang)
+                fp.close()
```
Idem.

tomaarsen commented on nltk/corpus/reader/extended_omw.py:
```diff
@@ -0,0 +1,141 @@
+                        self._lang_data[lang][index][offset_pos].add(val)
+
+    def disable_custom_lemmas(self, lang):
+        """prevent synsets from mistakenly added"""
```
Is this missing "being"?

tomaarsen reviewed (changes requested):
I recognise that this is a bit of an annoying situation, because any (automated) tests will fail due to `extended_omw` not yet existing in `nltk_data`, but perhaps you can provide a bit of sample code that should work if we (manually) download `extended_omw` and place it in our `nltk_data` folder?

Beyond that I added some small nitpicks. I can't really comment on the "real functionality", because I haven't been able to run it.

ExplorerFreda committed:
Change extended_omw reader w.r.t. review, add pre-commit

ExplorerFreda committed:
Merge branch 'develop' of https://github.com/ExplorerFreda/nltk into develop

* 'develop' of https://github.com/ExplorerFreda/nltk:
  Temporarily pause Python 3.10 CI tests due to scikit-learn issues with Windows
  Resolve IndexError in `sent_tokenize` (#2922)
  Drop support for Python 3.6, support Python 3.10 (#2920)
  updates for 3.6.6
  minor clean ups
  updates for 3.6.6

ExplorerFreda commented:
> I recognise that this is a bit of an annoying situation, because any (automated) tests will fail due to `extended_omw` not yet existing in `nltk_data`, but perhaps you can provide a bit of sample code that should work if we (manually) download `extended_omw` and place it in our `nltk_data` folder?
> 
> Beyond that I added some small nitpicks. I can't really comment on the "real functionality", because I haven't been able to run it.

Hi @tomaarsen, thanks for the review and suggestions! I've changed the code correspondingly. 

Below is the script for downloading the required data that should work: 
```bash
cd $NLTK_DATA_DIR/corpora/
wget http://compling.hss.ntu.edu.sg/omw/wn-wikt.zip -O extended_omw.zip
unzip extended_omw.zip
mv data/ extended_omw/
```

tomaarsen commented on nltk/corpus/reader/extended_omw.py:
```diff
@@ -0,0 +1,137 @@
+# Natural Language Toolkit: Exteneded Open Multilingual WordNet Reader
```
```suggestion
# Natural Language Toolkit: Extended Open Multilingual WordNet Reader
```
I haven't had the time to take a full look at this PR yet, but I noticed this small typo.

ExplorerFreda committed:
Update nltk/corpus/reader/extended_omw.py

fix typo

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

ekaf commented:
It would be nice to add the following capabilities:

1. Loading the English wiktionary wordnet. This requires small modifications in the wordnet module because it does not anticipate English wordnets from alternative provenances at this moment.

2. Loading different wordnets for the same language separetely would allow to compare the quality of different provenances. The reason why the 'wikt' and 'cldr' provenances did not go into the official OMW 1.4 package was that their quality was deemed uncertain. For each language, this PR uses the set union of all provenances, which may be appropriate for some uses, but it would be nice to also support different use cases.

3. Combining extended_wordnet with newer wordnet packages like wordnet31 and wordnet2021. This would be easier if the extended_omw corpus reader was a leaner class, that could be added to any English Wordnet, including the forthcoming ones like Open English Wordnet 2022.

4. For completeness, maybe also consider adding back the Persian wordnet and the alternative Chinese wordnet from the old OMW package (in the 'pas' and 'qcn' directories), which were left out of OMW 1.4 for quality reasons though it is possible that some users miss them.

stevenbird commented:
Resolved in #2946

stevenbird commented:
@ExplorerFreda thanks for your work, which is appreciated even though we ended up accepting a different PR.
